src/logic/hydra.lean

```diff
@@ -0,0 +1,181 @@
+arbitrary finite number of steps).
+
+This result is stated as the well-foundedness of the `cut_expand` relation defined in
+this file: we model the heads of the hydra as a multiset of elements of `α`, and the
+valid "moves" of the game are modelled by the relation `cut_expand r` on `multiset α`:
+`cut_expand r s' s` is true iff `s'` is obtained by removing one head `a ∈ s` and
+adding back an arbitrary multiset `t` of heads such that all `a' ∈ t` satisfy `r a' a`.
+
+To prove this theorem, we follow the proof by Peter LeFanu Lumsdaine at
+https://mathoverflow.net/a/229084/3332, and introduce the notion of `fibration` of relations, and
+a special addition of relations `game_add` that is used to define addition of games in
+combinatorial game theory.
+
+TODO: formalize the relations corresponding to more powerful (e.g. Kirby–Paris and Buchholz)
+hydras, and prove their well-foundedness.
+-/
+
+namespace relation
+
+variables {α β : Type*}
+
+section two_rels
+variables (rα : α → α → Prop) (rβ : β → β → Prop) (f : α → β)
+
+/-- A function `f : α → β` is a fibration between the relation `rα` and `rβ` if for all
+  `a : α` and `b : β`, whenever `b : β` and `f a` are related by `rβ`, `b` is the image
+  of some `a' : α` under `f`, and `a'` and `a` are related by `rα`. -/
+def fibration := ∀ ⦃a b⦄, rβ b (f a) → ∃ a', rα a' a ∧ f a' = b
+
+variables {rα rβ}
+/-- If `f : α → β` is a fibration between relations `rα` and `rβ`, and `a : α` is
+  accessible under `rα`, then `f a` is accessible under `rβ`. -/
+lemma _root_.acc.of_fibration (fib : fibration rα rβ f) {a} (ha : acc rα a) : acc rβ (f a) :=
+begin
+  induction ha with a ha ih,
+  refine acc.intro (f a) (λ b hr, _),
+  obtain ⟨a', hr', rfl⟩ := fib hr,
+  exact ih a' hr',
+end
+
+lemma _root_.acc.of_downward_closed (dc : ∀ {a b}, rβ b (f a) → b ∈ set.range f)
+  (a : α) (ha : acc (inv_image rβ f) a) : acc rβ (f a) :=
+ha.of_fibration f (λ a b h, let ⟨a', he⟩ := dc h in ⟨a', he.substr h, he⟩)
+
+variables (rα rβ)
+/-- The "addition of games" relation in combinatorial game theory, on the product type: if
+  `rα a' a` means that `a ⟶ a'` is a valid move in game `α`, and `rβ b' b` means that `b ⟶ b'`
+  is a valid move in game `β`, then `game_add rα rβ` specifies the valid moves in the juxtaposition
+  of `α` and `β`: the player is free to choose one of the games and make a move in it,
+  while leaving the other game unchanged. -/
+inductive game_add : α × β → α × β → Prop
+| fst {a' a b} : rα a' a → game_add (a',b) (a,b)
+| snd {a b' b} : rβ b' b → game_add (a,b') (a,b)
+
+/-- `game_add` is a `subrelation` of `prod.lex`. -/
+lemma game_add_le_lex : game_add rα rβ ≤ prod.lex rα rβ :=
+λ _ _ h, h.rec (λ _ _ b, prod.lex.left b b) (λ a _ _, prod.lex.right a)
+
+/-- `prod.rprod` is a subrelation of the transitive closure of `game_add`. -/
+lemma rprod_le_trans_gen_game_add : prod.rprod rα rβ ≤ trans_gen (game_add rα rβ) :=
+λ _ _ h, h.rec $ begin
```

Comment:
You don't need the `$`. Preferably this would all be in term mode, but I can't get that to work for some reason.

Reply:
Yes, I often see such unification issues with recursors and induction, e.g. the reason I need to split [line 82-83 here](https://gist.github.com/alreadydone/9eea203363bb76e5306c35a75550b836) into two `refine`s is probably due to such unification issue. Supplying more arguments to `game_add.fst` and `snd` might work but would make the proof ugly with lots of placeholders `_`.